src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -131,6 +131,372 @@ static void uct_ib_mlx5_parse_relaxed_order(uct_ib_mlx5_md_t *md,
+        status = UCS_OK;
+        goto out_destroy_dct_qp;
+    }
+
+    /* always set global address parameters, in case the port is RoCE or SRIOV */
+    attr.qp_state                  = IBV_QPS_RTR;
+    attr.min_rnr_timer             = 1;
+    attr.path_mtu                  = IBV_MTU_256;
+    attr.ah_attr.port_num          = 1;
+    attr.ah_attr.sl                = 0;
+    attr.ah_attr.is_global         = 1;
+    attr.ah_attr.grh.hop_limit     = 1;
+    attr.ah_attr.grh.traffic_class = 0;
+    attr.ah_attr.grh.sgid_index    = 0;
+
+    ret = ibv_modify_qp(dct_qp, &attr, IBV_QP_STATE |
+                                       IBV_QP_MIN_RNR_TIMER |
+                                       IBV_QP_AV |
+                                       IBV_QP_PATH_MTU);
+
+    if (ret != 0) {
+        ucs_debug("%s: failed to ibv_modify_qp(DCT, RTR): %m",
+                  uct_ib_device_name(dev));
+        status = UCS_OK;
+        goto out_destroy_dct_qp;
+    }
+
+    uct_ib_mlx5dv_qp_init_attr(&qp_init_attr, srq, pd, cq, IBV_QPT_DRIVER);
+    qp_init_attr.cap.max_recv_wr = 0;
+    uct_ib_mlx5_dci_qp_update_attr(&qp_init_attr);
+    uct_ib_mlx5dv_dc_qp_init_attr(&dv_attr, MLX5DV_DCTYPE_DCI);
+    dci_qp = UCS_PROFILE_CALL_ALWAYS(mlx5dv_create_qp, ctx, &qp_init_attr,
+                                     &dv_attr);
+    if (dci_qp == NULL) {
+        ucs_debug("%s: mlx5dv_create_qp(DCI) failed: %m",
+                  uct_ib_device_name(dev));
+        status = UCS_OK;
+        goto out_destroy_dct_qp;
+    }
+
+    status = uct_ib_mlx5dv_tx_wqe_ratio(dci_qp, &qp_init_attr,
+                                        &md->dv_tx_wqe_ratio.dc);
+    if (status != UCS_OK) {
+        goto out_destroy_dci_qp;
+    }
+
+    dev->flags |= UCT_IB_DEVICE_FLAG_DC;
+
+out_destroy_dci_qp:
+    uct_ib_destroy_qp(dci_qp);
+out_destroy_dct_qp:
+    uct_ib_destroy_qp(dct_qp);
+out_close_tmp_resources:
+    uct_ib_mlx5dv_close_tmp_resources(srq, pd, cq);
+out:
+#endif
+    if (!(dev->flags & UCT_IB_DEVICE_FLAG_DC)) {
+        md->dv_tx_wqe_ratio.dc = 0;
+    }
+
+    if (status == UCS_OK) {
+        ucs_debug("%s: DC %s supported", uct_ib_device_name(dev),
+                  (dev->flags & UCT_IB_DEVICE_FLAG_DC) ? "is" : "is not");
+    }
+    return status;
+}
+
+static ucs_status_t uct_ib_mlx5dv_obtain_ud_tx_wqe_ratio(uct_ib_mlx5_md_t *md)
+{
+    uct_ib_qp_init_attr_t qp_init_attr = {};
+    uct_ib_device_t *dev               = &md->super.dev;
+    ucs_status_t status;
+    struct ibv_qp *qp;
+    struct ibv_srq *srq;
+    struct ibv_pd *pd;
+    struct ibv_cq *cq;
+
+    status = uct_ib_mlx5dv_create_tmp_resources(dev, &srq, &pd, &cq);
+    if (status != UCS_OK) {
+        goto out;
+    }
+
+    uct_ib_mlx5dv_qp_init_attr(&qp_init_attr, srq, pd, cq, IBV_QPT_UD);
+    qp_init_attr.cap.max_recv_wr     = 128;
+    qp_init_attr.cap.max_send_sge    = 2;
+    qp_init_attr.cap.max_recv_sge    = 1;
+    qp_init_attr.cap.max_inline_data = 0;
```

Comment:
1. how can we make sure these values are in sync with ud transport?
2. seems wrong to have so much transport specific code in the common md. maybe we can initialize the wqe ratio in md struct to 0, and the transports will set them on first use?